File: src/test/java/com/runninglane/dto/buddy/test/java/cases/notnull/NonNullableTest.java
package com.runninglane.dto.buddy.test.java.cases.notnull;

import com.runninglane.dto.buddy.javainterop.DtoBuddy;
import com.runninglane.dto.buddy.test.NamingStrategyWithCountUpSuffix;
import org.junit.jupiter.api.Test;

import java.lang.annotation.Annotation;
import java.util.Arrays;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

public class NonNullableTest {
    private final DtoBuddy dtoBuddy = new DtoBuddy(new NamingStrategyWithCountUpSuffix());

    @Test
    public void testInterfaceAndGetter() throws NoSuchMethodException {
        Class<?> concreteClass = dtoBuddy.implement(DtoForCaseInterfaceAndGetter.class);
        DtoForCaseInterfaceAndGetter dto = dtoBuddy.create(
            concreteClass,
            Map.of("notNullable", new Dummy("Not Nullable Dummy"))
        );
        assertNotNull(dto);
        assertEquals("Not Nullable Dummy", dto.getNotNullable().getName());
        assertNull(dto.getNullable());
        assertTrue(
            Arrays.stream(concreteClass.getMethod("getNotNullable").getAnnotations())
                .anyMatch((annotation) -> annotation.toString().contains("NotNull"))
        );
        assertFalse(
            Arrays.stream(concreteClass.getMethod("getNullable").getAnnotations())
                .anyMatch((annotation) -> annotation.toString().contains("NotNull"))
        );
    }

    @Test
    public void testAbstractClassAndAbstractGetter() throws NoSuchMethodException {
        Class<?> concreteClass = dtoBuddy.implement(DtoForCaseAbstractClassAndAbstractGetter.class);
        DtoForCaseAbstractClassAndAbstractGetter dto = dtoBuddy.create(
            concreteClass,
            Map.of("notNullable", new Dummy("Not Nullable Dummy"))
        );
        assertNotNull(dto);
        assertEquals("Not Nullable Dummy", dto.getNotNullable().getName());
        assertNull(dto.getNullable());
        assertTrue(
            Arrays.stream(concreteClass.getMethod("getNotNullable").getAnnotations())
                .anyMatch((annotation) -> annotation.toString().contains("NotNull"))
        );
        assertFalse(
            Arrays.stream(concreteClass.getMethod("getNullable").getAnnotations())
                .anyMatch((annotation) -> annotation.toString().contains("NotNull"))
        );
    }
}

File: src/test/kotlin/com/runninglane/dto/buddy/test/cases/notnull/NonNullableTest.kt
package com.runninglane.dto.buddy.test.cases.notnull

import com.runninglane.dto.buddy.DtoBuddy
import com.runninglane.dto.buddy.test.NamingStrategyWithCountUpSuffix
import kotlin.test.*

class NonNullableTest {
    private val dtoBuddy = DtoBuddy(NamingStrategyWithCountUpSuffix())

    @Test
    fun testInterfaceAndProperty() {
        val concreteClass = dtoBuddy.implement(DtoForCaseInterfaceAndProperty::class)
        val dto: DtoForCaseInterfaceAndProperty = dtoBuddy.create(
            concreteClass,
            mapOf("notNullable" to Dummy("Not Nullable Dummy"))
        )
        assertNotNull(dto)
        assertEquals("Not Nullable Dummy", dto.notNullable.name)
        assertNull(dto.nullable)
    }

    @Test
    fun testInterfaceAndGetter() {
        val concreteClass = dtoBuddy.implement(DtoForCaseInterfaceAndGetter::class)
        val dto: DtoForCaseInterfaceAndGetter = dtoBuddy.create(
            concreteClass,
            mapOf("notNullable" to Dummy("Not Nullable Dummy"))
        )
        assertNotNull(dto)
        assertEquals("Not Nullable Dummy", dto.getNotNullable().name)
        assertNull(dto.getNullable())
    }

    @Test
    fun testAbstractClassAndAbstractProperty() {
        val concreteClass = dtoBuddy.implement(DtoForCaseAbstractClassAndAbstractProperty::class)
        val dto: DtoForCaseAbstractClassAndAbstractProperty = dtoBuddy.create(
            concreteClass,
            mapOf("notNullable" to Dummy("Not Nullable Dummy"))
        )
        assertNotNull(dto)
        assertEquals("Not Nullable Dummy", dto.notNullable.name)
        assertNull(dto.nullable)
    }

    @Test
    fun testAbstractClassAndAbstractGetter() {
        val concreteClass = dtoBuddy.implement(DtoForCaseAbstractClassAndAbstractGetter::class)
        val dto: DtoForCaseAbstractClassAndAbstractGetter = dtoBuddy.create(
            concreteClass,
            mapOf("notNullable" to Dummy("Not Nullable Dummy"))
        )
        assertNotNull(dto)
        assertEquals("Not Nullable Dummy", dto.getNotNullable().name)
        assertNull(dto.getNullable())
    }
}